# Segmenting and Clustering Neighborhoods in Toronto

In this assignment, we are going to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas  dataframe, explore, segment, and cluster the neighborhoods in the city of Toronto based on the postalcode and borough information.

This notebook contains all three questions of the assignment.

## Part 1

Import libraries.

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

Scrape data using BeautifulSoup.

In [2]:
postal_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(postal_url).text
soup  = BeautifulSoup(source, 'xml')

Ignore lines with a 'Not assigned' Borough, create pandas dataframe, clean up borough names and combine neighborhoods by postal codes. 

In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

We are going to assign borough name to all "Not assigned" neighbourhoods.

In [4]:
df['Neighborhood'][df['Neighborhood'] == 'Not assigned'] = df['Borough'][df['Neighborhood'] == 'Not assigned']
df.Neighborhood.str.count("Not assigned").sum()

0

In [5]:
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [6]:
df.shape

(103, 3)

Question 1 answer: We have 103 rows in the dataframe.

## Part 2

In [8]:
#!pip install geocoder

We are going to get the latitude and longitude coordinates for each neighborhood using geocoder.

The geocoder approach above took too long, so we had to stop it. As suggested, we are going to import the CSV file from url.

In [9]:
geo_data = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
geo_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Let's check dimensions and types of geo_data and df dataframes and merge them together.

In [10]:
print("The shape of borough and neighborhood data is", df.shape, "and the shape of latitude and longitude data is", geo_data.shape)

The shape of borough and neighborhood data is (103, 3) and the shape of latitude and longitude data is (103, 3)


In [11]:
print("The types of columns in borough and neighborhood data are: \n", df.dtypes, "\n and the types of columns in latitude and longitude data are: \n", geo_data.dtypes)

The types of columns in borough and neighborhood data are: 
 PostalCode      object
Borough         object
Neighborhood    object
dtype: object 
 and the types of columns in latitude and longitude data are: 
 Postal Code     object
Latitude       float64
Longitude      float64
dtype: object


In [12]:
geo_data = geo_data.rename(columns={'Postal Code': 'PostalCode'})
geo_data

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [13]:
comb_data = df.merge(geo_data.set_index('PostalCode'), on='PostalCode', how='inner')
comb_data

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

The dataframe looks as expected, has 103 rows and 5 columns, so we have a good combined dataframe.

## Part 3

In this section, we are going to Explore and cluster the neighborhoods in Toronto. We are going to replicate the same analysis we ddid for NYC data using Kmeans and Foursquare API.

In [19]:
#!conda install -c conda-forge geopy --yes

In [20]:
#!conda install -c conda-forge folium=0.5.0 --yes

In [14]:
import geocoder
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors
from sklearn.cluster import KMeans # import k-means from clustering stage
import folium # map rendering library

In [15]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(comb_data['Borough'].unique()),
        comb_data.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


We are going to use geopy library to get the latitude and longitude values of Totonto.
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>toronto_explorer</em>, as shown below.

In [16]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Create a map of Totonto with neighborhoods superimposed on top.

In [21]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(comb_data['Latitude'], comb_data['Longitude'], comb_data['Borough'], comb_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Define Foursquare Credentials and Version

In [22]:
CLIENT_ID = 'LBUSXEOOYJPIKVGEWITLGEII3AOXIBZEL1YZCX1B00KARGYI' # your Foursquare ID
CLIENT_SECRET = 'X0JBN03SNTXKMAV0ADFGLSK1X42GOF0DPZV20VCT41EVFUXU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

We are going to create a function to get all the venue categories in Toronto

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'],
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue',
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Next, we are going to collect the venues in Toronto for each neighborhood and create a new dataframe caled _toronto_venues_.

In [25]:
toronto_venues = getNearbyVenues(comb_data['Neighborhood'], 
                                 comb_data['Latitude'], 
                                 comb_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [26]:
toronto_venues.shape

(1337, 7)

So we have 1337 records and 7 columns. Let's check the resulting dataframe.

In [27]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1   Fast Food Restaurant  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

Let's check the number of venues returned for each neighborhood.

In [28]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    25   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                3   
Woodbine Heights                                                      6   
York Mills West                                                       3   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              4   
Alderwood, Long Branch                                                 8   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     25   
...                                                                  ...   
Willowdale West                                                        5   
Willowdale, Newtonbrook                                                1   
Woburn                                                                 3   
Woodbine Heights                                                       6   
York Mills West                                                        3   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   25              25   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               3               3   
Woodbine Heights                                     6               6   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             25   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         3   
Woodbine Heights                                               6   
York Mills West                                                3   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     4  
Alderwood, Long Branch                

In [29]:
print('There are {} uniques venue categories in Toronto.'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 uniques venue categories in Toronto.


### We are now going to analyze each neighbourhood.

In [30]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Truck Stop  \
0                 0                    0  ...           0   
1                 0                    0  ...           0   
2                 0                    0  ...           0   
3                 0                    0  ...           0   
4                 0                    0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 237 columns]

In [31]:
print('The new dataframe size is {}.'.format(toronto_onehot.shape))

The new dataframe size is (1337, 237).


We are now going to group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category.

In [32]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
95                                  Willowdale West                0.0   
96                          Willowdale, Newtonbrook                0.0   
97                                           Woburn                0.0   
98                                 Woodbine Heights                0.0   
99                                  York Mills West                0.0   

    Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0              0.0      0.0                 0.0           0.0             0.0   
1              0.0      0.0                 0.0           0.0             0.0   
2              0.0      0.0                 0.0           0.0             0.0   
3              0.0      0.0                 0.0           0.0             0.0   
4              0.0      0.0                 0.0           0.0             0.0   
..             ...      ...                 ...           ...             ...   
95             0.0      0.0                 0.0           0.0             0.0   
96             0.0      0.0                 0.0           0.0             0.0   
97             0.0      0.0                 0.0           0.0             0.0   
98             0.0      0.0                 0.0           0.0             0.0   
99             0.0      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  ...  Truck Stop  \
0               0.0               0.0                 0.00  ...         0.0   
1               0.0               0.0                 0.00  ...         0.0   
2               0.0               0.0                 0.00  ...         0.0   
3               0.0               0.0                 0.00  ...         0.0   
4               0.0               0.0                 0.04  ...         0.0   
..              ...               ...                  ...  ...         ...   
95              0.0               0.0                 0.00  ...         0.0   
96              0.0               0.0                 0.00  ...         0.0   
97              0.0               0.0                 0.00  ...         0.0   
98              0.0               0.0                 0.00  ...         0.0   
99              0.0               0.0                 0.00  ...         0.0   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                             0.0               0.0     0.000000   
1                             0.0               0.0     0.000000   
2                             0.0               0.0     0.000000   
3                             0.0               0.0     0.000000   
4                             0.0               0.0     0.000000   
..                            ...               ...          ...   
95                            0.0               0.0     0.000000   
96                            0.0               0.0     0.000000   
97                            0.0               0.0     0.000000   
98                            0.0               0.0     0.166667   
99                            0.0               0.0     0.000000   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                     0.0              0.0       0.0          0.0   
1                     0.0              0.0       0.0          0.0   
2                     0.0              0.0       0.0          0.0   
3                     0.0              0.0       0.0          0.0   
4                     0.0              0.0       0.0          0.0

In [33]:
print('And the new size is {}.'.format(toronto_grouped.shape))

And the new size is (100, 237).


Let's write a function to sort the venues in descending order and choose the top 10 for each neighbourhood.

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [35]:
import numpy as np # library to handle data in a vectorized manner

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East            Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                Coffee Shop          Dance Studio        Sandwich Place   
2                       Bank     Mobile Phone Shop           Bridal Shop   
3                       Bank   Japanese Restaurant                  Café   
4                Coffee Shop        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Yoga Studio             Drugstore                   Donut Shop   
1                   Pub                  Pool                          Gym   
2        Sandwich Place         Deli / Bodega                   Restaurant   
3           Yoga Studio         Deli / Bodega  Eastern European Restaurant   
4      Greek Restaurant     Indian Restaurant         Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0               Dog Run   Distribution Center             Discount Store  
1        Discount Store                 Diner         Dim Sum Restaurant  
2           Supermarket      Sushi Restaurant  Middle Eastern Restaurant  
3             Drugstore            Donut Shop                    Dog Run  
4                   Pub      Sushi Restaurant                    Butcher

### Now, we are going to create the model to cluster our neighbourhoods.

Run _k_-means to cluster the neighborhood into 5 clusters.

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [38]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 3, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1,
       3, 1, 1, 1, 1, 4, 1, 4, 1, 1, 1, 4, 1, 4, 1, 1, 1, 1, 3, 1, 1, 1,
       4, 1, 1, 4, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1,
       1, 1, 2, 1, 4, 1, 1, 1, 4, 1, 1, 4], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood by merging toronto_grouped with comb_data on neighbourhood to add latitude & longitude for each neighbourhood to prepare it for plotting.

In [39]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = comb_data

In [40]:
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head() # check the last columns!

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             4.0                  Park    Fast Food Restaurant   
1 -79.315572             1.0           Coffee Shop   Portuguese Restaurant   
2 -79.360636             1.0           Coffee Shop                  Bakery   
3 -79.464763             1.0        Clothing Store  Furniture / Home Store   
4 -79.389494             1.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio           Curling Ice   
1          Hockey Arena     French Restaurant               Dog Run   
2                  Park        Breakfast Spot               Theater   
3     Accessories Store              Boutique    Miscellaneous Shop   
4                   Bar              Beer Bar                   Spa   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0            Donut Shop               Dog Run    Distribution Center   
1   Distribution Center        Discount Store             Donut Shop   
2            Restaurant                  Café                    Pub   
3           Coffee Shop         Women's Store  Vietnamese Restaurant   
4         Smoothie Shop        Sandwich Place            Salad Place   

  9th Most Common Venue       10th Most Common Venue  
0        Discount Store                        Diner  
1      Cuban Restaurant           Dim Sum Restaurant  
2        Chocolate Shop                  Yoga Studio  
3     Electronics Store  Eastern European Restaurant  
4            Restaurant                Burrito Place

Let's drop all the NaN values for Cluster Labels.

In [41]:
toronto_merged = toronto_merged.dropna(subset=['Cluster Labels'])

Let's visualize the resulting clusters.

In [42]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(int(cluster)+1), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters
We are now going to examine each cluster and determine the discriminating venue categories that distinguish each cluster.

Cluster 1

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
25  Downtown Toronto             0.0         Grocery Store   
46        North York             0.0         Grocery Store   
60        North York             0.0          Liquor Store   
63              York             0.0         Grocery Store   
89         Etobicoke             0.0         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
25                  Café                  Park            Restaurant   
46                  Park         Shopping Mall                  Bank   
60         Grocery Store    Athletics & Sports  Gym / Fitness Center   
63        Breakfast Spot     Convenience Store              Bus Line   
89  Fast Food Restaurant              Pharmacy           Pizza Place   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
25    Athletics & Sports                  Candy Store    Italian Restaurant   
46           Curling Ice                    Drugstore            Donut Shop   
60        Discount Store             Department Store          Dessert Shop   
63         Deli / Bodega  Eastern European Restaurant             Drugstore   
89   Fried Chicken Joint                   Beer Store        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
25            Baby Store             Nightclub            Coffee Shop  
46               Dog Run   Distribution Center         Discount Store  
60    Dim Sum Restaurant                 Diner                Dog Run  
63            Donut Shop               Dog Run    Distribution Center  
89           Yoga Studio         Deli / Bodega       Department Store

Cluster 2

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York             1.0           Coffee Shop   
2         Downtown Toronto             1.0           Coffee Shop   
3               North York             1.0        Clothing Store   
4             Queen's Park             1.0           Coffee Shop   
6              Scarborough             1.0  Fast Food Restaurant   
..                     ...             ...                   ...   
97        Downtown Toronto             1.0                  Café   
98               Etobicoke             1.0                  Park   
99        Downtown Toronto             1.0      Sushi Restaurant   
100  East Toronto Business             1.0    Light Rail Station   
102              Etobicoke             1.0         Tanning Salon   

      2nd Most Common Venue 3rd Most Common Venue  \
1     Portuguese Restaurant          Hockey Arena   
2                    Bakery                  Park   
3    Furniture / Home Store     Accessories Store   
4          Sushi Restaurant                   Bar   
6               Yoga Studio           Escape Room   
..                      ...                   ...   
97              Coffee Shop            Restaurant   
98                    River            Smoke Shop   
99         Ramen Restaurant          Dance Studio   
100              Skate Park           Pizza Place   
102            Burger Joint       Supplement Shop   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1              French Restaurant               Dog Run   Distribution Center   
2                 Breakfast Spot               Theater            Restaurant   
3                       Boutique    Miscellaneous Shop           Coffee Shop   
4                       Beer Bar                   Spa         Smoothie Shop   
6    Eastern European Restaurant             Drugstore            Donut Shop   
..                           ...                   ...                   ...   
97            Seafood Restaurant   Japanese Restaurant             Gastropub   
98                          Pool            Donut Shop               Dog Run   
99          Ethiopian Restaurant           Escape Room        Ice Cream Shop   
100         Fast Food Restaurant        Farmers Market         Auto Workshop   
102                       Bakery     Convenience Store         Burrito Place   

    7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
1          Discount Store             Donut Shop      Cuban Restaurant   
2                    Café                    Pub        Chocolate Shop   
3           Women's Store  Vietnamese Restaurant     Electronics Store   
4          Sandwich Place            Salad Place            Restaurant   
6                 Dog Run    Distribution Center        Discount Store   
..                    ...                    ...                   ...   
97                    Pub              Speakeasy                   Bar   
98    Distribution Center         Discount Store                 Diner   
99      Indian Restaurant                    Pub              Beer Bar   
100         Burrito Place                    Spa            Comic Shop   
102  Fast Food Restaurant         Hardware Store        Sandwich Place   

          10th Most Common Venue  
1             Dim Sum Restaurant  
2                    Yoga Studio  
3    Eastern European Restaurant  
4                  Burrito Place  
6                          Diner  
..                           ...  
97                    Steakhouse  
98            Dim Sum Restaurant  
99            Italian Restaurant  
100                      Brewery  
102                          Gym  

[79 rows x 12 columns]

Cluster 3

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             2.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
11   Distribution Center         Discount Store

Cluster 4

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
53   North York             3.0      Business Service        Baseball Field   
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0        Baseball Field      Business Service   

    3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
53             Food Truck           Yoga Studio  Eastern European Restaurant   
57   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
101           Yoga Studio           Escape Room  Eastern European Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
53              Drugstore            Donut Shop               Dog Run   
57              Drugstore            Donut Shop               Dog Run   
101             Drugstore            Donut Shop               Dog Run   

    9th Most Common Venue 10th Most Common Venue  
53    Distribution Center         Discount Store  
57    Distribution Center         Discount Store  
101   Distribution Center         Discount Store

Cluster 5

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             4.0                  Park   
10              North York             4.0                  Park   
21                    York             4.0                  Park   
35  East York/East Toronto             4.0                  Park   
52              North York             4.0                  Park   
61         Central Toronto             4.0                  Park   
64                    York             4.0                  Park   
66              North York             4.0                  Park   
77               Etobicoke             4.0                  Park   
83         Central Toronto             4.0                  Park   
85             Scarborough             4.0                  Park   
91        Downtown Toronto             4.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Fast Food Restaurant     Food & Drink Shop           Yoga Studio   
10                  Playground   Japanese Restaurant                Bakery   
21               Women's Store                  Pool      Cuban Restaurant   
35           Convenience Store           Yoga Studio           Curling Ice   
52                 Yoga Studio      Cuban Restaurant             Drugstore   
61                 Swim School              Bus Line     Electronics Store   
64               Jewelry Store           Yoga Studio          Dance Studio   
66  Construction & Landscaping     Convenience Store           Yoga Studio   
77              Sandwich Place           Yoga Studio           Curling Ice   
83                  Playground                 Trail              Creperie   
85                  Playground          Intersection           Yoga Studio   
91                  Playground                 Trail              Creperie   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Curling Ice            Donut Shop               Dog Run   
10                  Curling Ice             Drugstore            Donut Shop   
21                    Drugstore            Donut Shop               Dog Run   
35                    Drugstore            Donut Shop               Dog Run   
52                   Donut Shop               Dog Run   Distribution Center   
61                    Drugstore            Donut Shop               Dog Run   
64  Eastern European Restaurant             Drugstore            Donut Shop   
66                  Curling Ice             Drugstore            Donut Shop   
77                    Drugstore            Donut Shop               Dog Run   
83                   Donut Shop               Dog Run   Distribution Center   
85             Cuban Restaurant            Donut Shop               Dog Run   
91                   Donut Shop               Dog Run   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Distribution Center        Discount Store                  Diner  
10               Dog Run   Distribution Center         Discount Store  
21   Distribution Center        Discount Store                  Diner  
35   Distribution Center        Discount Store                  Diner  
52        Discount Store                 Diner     Dim Sum Restaurant  
61   Distribution Center        Discount Store                  Diner  
64               Dog Run   Distribution Center         Discount Store  
66               Dog Run   Distribution Center         Discount Store  
77   Distribution Center        Discount Store                  Diner  
83        Discount Store                 Diner     Dim Sum Restaurant  
85   Distribution Center        Discount Store                  Diner  
91        Discount Store                 Diner     Dim Sum Restaurant

We have successfully clustered Toronto neighbourhoods by venue categories:
- Cluster 1: Grocery stores
- Cluster 2: Food (coffee shops, restaurants, donut shops etc.)
- Cluster 3: Bakery and yoga
- Cluster 4: Baseball
- Cluster 5: Outside fun (parks, playgrounds, trail, dog run etc.)